### ISSUE TITLE
Why do I get this error message when I'm trying to save an artifact?

### ISSUE THREAD

**Issue by cqzhy1986** [salimfadhley/jenkinsapi#403]
Hi guys,
I'm using the latest Jenkins 1.637 to create a software-project-building job for my company.
After the job finished, I use the tar tool to pack some libs into a tar.gz package and try to get it from user side via the 0.2.29 version of JenkinsAPI.
But I will get, for every time, an error message from the JenkinsAPI's Artifact.save() function:
"ERROR:root:Failed request at http://xxx.xxx.xxx.xxx:xxxxxx/fingerprint/xxxxxxxxxxxxxxxx/api/python with params: None"
Finally I can still find that the artifact has been saved to the path I specified.
Just curious why I always get this error mesasge.
Could someone please help me?

**Comment by ripcurld00d** [comment]
@cqzhy1986 - My guess is that there is no fingerprint set for this artifact, @lechat correct me if I'm wrong.

To fix this, try the following:
* Go to the "Configuration" page of that specific job
* Add the "Record fingerprints of files to track usage" post-build action at the `Post-build action` section
* Click on the save button

**Comment by cqzhy1986** [comment]
@ripcurld00d - Thank you. Your solution can fix it!

**Comment by lechat** [comment]
I'll leave this issue open as it will be good to fail with meaningfull message, this shall be fixed.

**Comment by ripcurld00d** [comment]
@lechat :+1: 
What I suggest is to add [here](https://github.com/salimfadhley/jenkinsapi/blob/master/jenkinsapi/fingerprint.py#L59) a call to `logging.info` saying "could not find fingerprint" with a link to this issue or to a document explaining how to add fingerprints to artifacts in Jenkins.

This should clearify the `ERROR:root:Failed request at http://xxx.xxx.xxx.xxx:xxxxxx/fingerprint/xxxxxxxxxxxxxxxx/api/python with params: None` message.

Thoughts?

**Comment by lechat** [comment]
It shall be ``logging.warning`` actually. I think that the error happens in [Fingerprint.\_\_init\_\_:29](https://github.com/salimfadhley/jenkinsapi/blob/master/jenkinsapi/fingerprint.py#L29)

@cqzhy1986, could you give us stacktrace for this error?

**Comment by cqzhy1986** [comment]
@lechat , what I was doing was quite simple:
```
def get_artifact(url, job, build_num, artifact_name, user_name, passwd, local_path):
  art_dict = jenkinsapi.api.get_artifacts(url, job, build_num, user_name, passwd)
  art = art_dict[artifact_name]
  art.save(local_path)
```
Is this information helpful to you? This is the only code I'm using JenkinsAPI in my script.

**Comment by ripcurld00d** [comment]
@cqzhy1986 -- sorry but no. I need to see the entire error message you receive when you execute this code and there's no fingerprints set for the artifacts.

Something similar to this:

>Traceback (most recent call last):
>         File "<doctest...>", line 10, in <module>
>        lumberjack()
>       File "<doctest...>", line 4, in lumberjack
>    bright_side_of_death()
>IndexError: tuple index out of range

**Comment by cqzhy1986** [comment]
@ripcurld00d , but this is the only error message I've got. Although the error message was printed by JenkinsAPI, the program can still continue running and finish successfully, so there's no such a "Traceback".

**Comment by ripcurld00d** [comment]
@cqzhy1986 -- thanks anyway.

@lechat -- I have successfully reproduced it. You were right, it failed where you said.

```
INFO:jenkinsapi.artifact:Saving artifact @ http://172.17.0.1:8080/job/my_files/1/artifact/datefile.txt to /tmp/bla2.txt
WARNING:jenkinsapi.artifact:Attempt to change the filename of artifact datefile.txt on save.
INFO:jenkinsapi.artifact:Local file is missing, downloading new.
INFO:root:in fingerprint with url 'http://172.17.0.1:8080/fingerprint/fd0baddf7e5826179e2e3cc37e7378a5/'
INFO:root:@poll 'http://172.17.0.1:8080/fingerprint/fd0baddf7e5826179e2e3cc37e7378a5'
INFO:root:@_poll 'http://172.17.0.1:8080/fingerprint/fd0baddf7e5826179e2e3cc37e7378a5'
ERROR:root:status code 404
ERROR:root:Failed request at http://172.17.0.1:8080/fingerprint/fd0baddf7e5826179e2e3cc37e7378a5/api/python with params: None 
iNFO:jenkinsapi.fingerprint:@valid
```

What do you think is the right way to handle this.
1. We can set in the Artifact class an attribute that determines whether to check fingerprints
2. Run this check before initializing the `Fingerprint` instance

**Comment by lechat** [comment]
The fingerprint in this case is used to check md5 of downloaded file against md5 generated by Jenkins (fingerprint). I think that the best approach is to catch that error in Artifact.save, log warning that md5 cannot be checked if fingerprints are not enabled and still let saving of the artifact w/o md5 check.

I think that will be less confusing to people, who don't use fingerprints.

**Comment by ripcurld00d** [comment]
@lechat I can write a patch that resolves this, if you don't mind.

**Comment by ripcurld00d** [comment]
@lechat, can you close this issue? it was resolved in #410. thanks
